addons/account/models/account_move.py

```diff
@@ -3205,70 +3210,85 @@ def _get_invoice_counterpart_amls_for_early_payment_discount_per_payment_term_li
         """
         self.ensure_one()
 
-        def grouping_key_generator(base_line, tax_values):
-            return self.env['account.tax']._get_generation_dict_from_base_line(base_line, tax_values)
-
         def inverse_tax_rep(tax_rep):
             tax = tax_rep.tax_id
             index = list(tax.invoice_repartition_line_ids).index(tax_rep)
             return tax.refund_repartition_line_ids[index]
 
+        company = self.company_id
+        payment_term_line = self.line_ids.filtered(lambda x: x.display_type == 'payment_term')
+        tax_lines = self.line_ids.filtered(lambda x: x.display_type == 'tax')
+        invoice_lines = self.line_ids.filtered(lambda x: x.display_type == 'product')
+        payment_term = self.invoice_payment_term_id
+        early_pay_discount_computation = payment_term.early_pay_discount_computation
+        discount_percentage = payment_term.discount_percentage
+
+        res = {
+            'term_lines': defaultdict(lambda: {}),
+            'tax_lines': defaultdict(lambda: {}),
+            'base_lines': defaultdict(lambda: {}),
+        }
+        if not discount_percentage:
+            return res
+
         # Get the current tax amounts in the current invoice.
         tax_amounts = {
             inverse_tax_rep(line.tax_repartition_line_id).id: {
                 'amount_currency': line.amount_currency,
                 'balance': line.balance,
             }
-            for line in self.line_ids.filtered(lambda x: x.display_type == 'tax')
+            for line in tax_lines
         }
 
-        product_lines = self.line_ids.filtered(lambda x: x.display_type == 'product')
         base_lines = [
             {
                 **x._convert_to_tax_base_line_dict(),
                 'is_refund': True,
             }
-            for x in product_lines
+            for x in invoice_lines
         ]
         for base_line in base_lines:
             base_line['taxes'] = base_line['taxes'].filtered(lambda t: t.amount_type != 'fixed')
 
+            if early_pay_discount_computation == 'included':
+                remaining_part_to_consider = (100 - discount_percentage) / 100.0
+                base_line['price_unit'] *= remaining_part_to_consider
+
+        # Prepare the tax details for each line.
+        to_process = []
+        for base_line in base_lines:
+            tax_details_results = self.env['account.tax']._prepare_base_line_tax_details(
+                base_line,
+                company,
+                split_repartition_lines=True,
+            )
+            to_process.append((base_line, tax_details_results))
+
+        # Aggregate taxes.
+        def grouping_key_generator(base_line, tax_values):
+            return self.env['account.tax']._get_generation_dict_from_base_line(base_line, tax_values)
```

Comment:
bof, that's just a personal opinion but I like inner methods to be defined at the beginning of the method body. 

I actually wonder if that couldn't be removed, and set as the default deeper in the stacktrace?

Reply:
@qdp-odoo I have some work to do on the grouping key for multiple reasons:
- I have to align the tax totals python-side with the POS
- the grouping key is duplicated since the refactoring of account.move :robot: 
- this is currently the biggest problem to solve using round_globally

I plan to cleanup that on the next iteration to fix the round_globally